File: 135.candy.rb
#
# @lc app=leetcode id=135 lang=ruby
#
# [135] Candy
#

# @lc code=start
# @param {Integer[]} ratings
# @return {Integer}
def candy(ratings)
  sum = 1
  cons_inc = 0
  cons_dec = 0
  peak = 0

  (1...ratings.length).each do |i|
    if ratings[i] > ratings[i - 1]
      cons_dec = 0
      cons_inc += 1
      peak = cons_inc
      sum += cons_inc + 1
    elsif ratings[i] == ratings[i - 1]
      cons_dec = 0
      cons_inc = 0
      peak = 0
      sum += 1
    else
      cons_inc = 0
      cons_dec += 1
      sum += cons_dec + ((peak >= cons_dec) ? 0 : 1)
    end
  end

  sum
end
# @lc code=end

File: 135.candy.ts
/*
 * @lc app=leetcode id=135 lang=typescript
 *
 * [135] Candy
 */

// @lc code=start
function candy(ratings: number[]): number {
  let sum = 1;
  let consInc = 0;
  let consDec = 0;
  let peak = 0;

  for (let i = 1; i < ratings.length; i++) {
    if (ratings[i] > ratings[i - 1]) {
      consDec = 0;
      consInc++;
      peak = consInc;
      sum += consInc + 1;
    } else if (ratings[i] == ratings[i - 1]) {
      consDec = 0;
      consInc = 0;
      peak = 0;
      sum++;
    } else {
      consInc = 0;
      consDec++;
      sum += consDec + (peak >= consDec ? 0 : 1);
    }
  }

  return sum;
}
// @lc code=end
